### DISCUSSION TITLE
[HELP] Apache reverse proxy

### DISCUSSION THREAD

**Discussion by JHthe4** [jonaswinkler/paperless-ng#1482] [Support] [answered]
Hi all,

I've installed paperless-ng using the docker method, and it works like a charm when accessing it directly on its port. I would like to be able to access it via a sublocation on port 80 as well, where I have an apache2 server running (which I unfortunately can't change to nginx or something else). 

I've managed to configure the reverse proxy so that it forwards requests correctly, but yet all I receive is a white page which states "Paperless-ng is loading...". Below is my Location configuration for apache. I have enabled the modules proxy, proxy_http, proxy_wstunnel, headers and rewrite. 

```
<Location "/paperless">
                RewriteEngine On
                RewriteCond %{HTTP:Upgrade} websocket [NC]
                RewriteCond %{HTTP:Connection} upgrade [NC]
                RewriteRule ^/?(.*) "ws://127.0.0.1:8001/$1" [P,L]
                ProxyPass "http://127.0.0.1:8001/"
                ProxyPassReverse "http://127.0.0.1:8001/"
                ProxyPreserveHost On
                RequestHeader set X-Real-IP %{REMOTE_ADDR}s
</Location>
```

Maybe someone has an idea what I'm doing wrong. I'd be thankful for any help.

Cheers

**Comment by TheNomad11** [comment, marked as answer]
I am a noob, but mine works fine although it is much more basic. As far as I know, everybody is recommending Nginx

```
LoadModule proxy_module modules/mod_proxy.so
LoadModule proxy_http_module modules/mod_proxy_http.so
LoadModule proxy_wstunnel_module modules/mod_proxy_wstunnel.so
<IfModule mod_ssl.c>
<VirtualHost *:443>
    ServerName paperless.XXXXXX.com
    AllowEncodedSlashes On
    ProxyPreserveHost On
    ProxyPassMatch ^/(ws)$  ws://localhost:8888/$1
   ProxyPass / http://localhost:8888/
    
SSLCertificateFile /etc/letsencrypt/live/paperless.XXXXXXX.com/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/paperless.XXXXXX.com/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
</IfModule>

```

**Reply by JHthe4** [reply]
Thanks for the config! You're using a subdomain so you're on the same context right? This is where I'm failing right now. I just can't get the reverse translation to work, I'm redirected to XXX/accounts/login/?next= and not to XXX/paperless/accounts/login/?next. Maybe you have an idea how I can get the retranslation to work? Or is this a paperless-ng issue (with absolute link paths or something)?

**Reply by TheFehr** [reply]
Yes subfolder context is not supported at the moment, see #36 and #679

**Reply by sbrunner** [reply]
The reverse proxy didn't forward the original folder, then the only way to solve the issue is:
- make paperless-ng working on the subfolder /paperless
- make that the reverse proxy don't change the path

**Reply by JHthe4** [reply]
Thanks, you guys gave me the hints I needed to get it running. I assumed that the reverse proxy would take care of rewriting, but running it at a subpath fixes everything. All I had to do was set `PAPERLESS_FORCE_SCRIPT_NAME=/paperless` in my `docker-compose.env` and slightly modify my apache config: 
```
<Location "/paperless">
                ProxyPreserveHost On
                ProxyPass "http://localhost:8001/paperless"
                ProxyPassReverse "http://localhost:8001/paperless"
                RequestHeader set X-Real-IP %{REMOTE_ADDR}s
                RewriteEngine On
                RewriteCond %{HTTP:UPGRADE} ^WebSocket$ [NC]
                RewriteCond %{HTTP:CONNECTION} Upgrade$ [NC]
                RewriteRule ^/?(.*) "ws://127.0.0.1:8001/$1" [P,L]
                #Order allow,deny
                #Allow from all
</Location>
```

Right now it works perfectly like this. I'm not quite sure about the WebSocket rewriting (do I need a subpath here too?), but even that seems to work at the moment. Maybe this configuration might help someone else facing similar issues in the future.

Thanks again for all your help!

**Comment by egabosh** [comment]
Moin,
have no sublocation but had problems with csrftoken e.g while creating tags with apache reverse proxy.
Maybe this helps somebody:
https://github.com/jonaswinkler/paperless-ng/issues/1603
```
SetEnvIf Cookie "(^|;\ *)csrftoken=([^;\ ]+)" csrftoken=$2
RequestHeader set  X-CSRFToken "%{csrftoken}e"
ProxyPreserveHost On
AllowEncodedSlashes On
ProxyPass / http://127.0.0.1:8000/
ProxyPassMatch ^/(ws)$  ws://localhost:8000/$1
ProxyPassReverse / http://127.0.0.1:8000/
ProxyRequests off
```
Regards
olli
